mono/profiler/mono-profiler-log.c

```diff
@@ -5011,34 +4633,106 @@ mono_profiler_startup (const char *desc)
 	mono_profiler_install_thread (thread_start, thread_end);
+
+	//It's questionable whether we actually want this to be mandatory, maybe put it behind the actual event?
 	mono_profiler_install_thread_name (thread_name);
-	mono_profiler_install_enter_leave (method_enter, method_leave);
-	mono_profiler_install_jit_end (method_jitted);
-	mono_profiler_install_code_buffer_new (code_buffer_new);
-	mono_profiler_install_exception (throw_exc, method_exc_leave, clause_exc);
-	mono_profiler_install_monitor (monitor_event);
-	mono_profiler_install_runtime_initialized (runtime_initialized);
-	if (do_coverage)
+
+
+	if (config.effective_mask & DomainEvents) {
+		events |= MONO_PROFILE_APPDOMAIN_EVENTS;
+		mono_profiler_install_appdomain (NULL, domain_loaded, domain_unloaded, NULL);
+		mono_profiler_install_appdomain_name (domain_name);
+	}
+
+	if (config.effective_mask & AssemblyEvents) {
+		events |= MONO_PROFILE_ASSEMBLY_EVENTS;
+		mono_profiler_install_assembly (NULL, assembly_loaded, assembly_unloaded, NULL);
+	}
+
+	if (config.effective_mask & ModuleEvents) {
+		events |= MONO_PROFILE_MODULE_EVENTS;
+		mono_profiler_install_module (NULL, image_loaded, image_unloaded, NULL);
+	}
+
+	if (config.effective_mask & ClassEvents) {
+		events |= MONO_PROFILE_CLASS_EVENTS;
+		mono_profiler_install_class (NULL, class_loaded, class_unloaded, NULL);
+	}
+
+	if (config.effective_mask & JitCompilationEvents) {
+		events |= MONO_PROFILE_JIT_COMPILATION;
+		mono_profiler_install_jit_end (method_jitted);
+		mono_profiler_install_code_buffer_new (code_buffer_new);
+	}
+
+	if (config.effective_mask & ExceptionEvents) {
+		events |= MONO_PROFILE_EXCEPTIONS;
+		mono_profiler_install_exception (throw_exc, method_exc_leave, clause_exc);
+	}
+
+	if (config.effective_mask & AllocationEvents) {
+		events |= MONO_PROFILE_ALLOCATIONS;
+		mono_profiler_install_allocation (gc_alloc);
+	}
+
+	//GCEvents is mandatory
+	//ThreadEvents is mandatory
+
+	if (config.effective_mask & EnterLeaveEvents) {
+		events |= MONO_PROFILE_ENTER_LEAVE;
+		mono_profiler_install_enter_leave (method_enter, method_leave);
+	}
+
+	if (config.effective_mask & InsCoverageEvents) {
+		events |= MONO_PROFILE_INS_COVERAGE;
 		mono_profiler_install_coverage_filter (coverage_filter);
+	}
 
-	if (do_mono_sample && sample_freq) {
+	//XXX should we check for SamplingFeature instead??
+	if (config.effective_mask & SamplingEvents) {
 		events |= MONO_PROFILE_STATISTICAL;
-		mono_profiler_set_statistical_mode (sampling_mode, sample_freq);
+		mono_profiler_set_statistical_mode (config.sampling_mode, config.sample_freq);
 		mono_profiler_install_statistical (mono_sample_hit);
 	}
 
+	if (config.effective_mask & MonitorEvents) {
+		events |= MONO_PROFILE_MONITOR_EVENTS;
+		mono_profiler_install_monitor (monitor_event);
+	}
+
+	if (config.effective_mask & GCMoveEvents) {
+		events |= MONO_PROFILE_GC_MOVES;
+		mono_profiler_install_gc_moves (gc_moves);
+	}
+
+	if (config.effective_mask & (GCRootEvents | GCHandleEvents)) {
```

Comment:
Minor thing: It would be nice to add a `TODO` here since we should really split these events into two separate flags in the profiler API.